Improve changelog (#36)
Simple-Robotics/pinocchio: CHANGELOG.md

- Add `internal::matrix_product` in `math.hpp`
- Add `internal::matrix_inversion` in `math.hpp`
- Add `internal::matrix_inversion_code_generated` in `math.hpp`

### Changed

> - C++17 is minimal supported version

Reviewer: ```suggestion
- C++17 is now the minimal supported version of the C++ standard. Check [cppreference](https://en.cppreference.com/w/cpp/compiler_support/17) to see if your compiler supports it.
```

- Following algorithms are now unsupported algorithms (no more deprecated):
  - `forwardDynamics`
  - `impulseDynamics`
  - `getKKTContactDynamicMatrixInverse`
- Change arguments in `initConstraintDynamics`
- Change arguments in `BaumgarteCorrectorParametersTpl` constructor: scalar are used instead of vectors
- Deprecate `ContactCholeskyDecompositionTpl`, replaced by  `ConstraintCholeskyDecompositionTpl`
- Deprecate `DelassusCholeskyExpressionTpl`, replaced by  `DelassusOperatorCholeskyExpressionTpl`
- Major refactorization of `ConstraintCholeskyDecompositionTpl` to ease online resizing
- Deprecate `ConstraintCholeskyDecompositionTpl::allocate` replaced by `ConstraintCholeskyDecompositionTpl::rebuild`
- Change arguments in `ConstraintCholeskyDecompositionTpl` constructor

### Removed
- Remove unused headers `deprecated-macros.hpp` and `deprecated-namespaces.hpp`
- Remove header `pinocchio/deprecation.hpp`, directly use generated `pinocchio/deprecated.hpp`
- macros.hpp: remove macros already provided by jrl-cmakemodules
- Remove pinocchio 3 deprecated files and functions (see `doc/_porting/porting-3-to-4.md`)

> - Remove `PINOCCHIO_WITH_CXX{11,14,17}_SUPPORT` define

Reviewer: ```suggestion
- Remove `PINOCCHIO_WITH_CXX{11,14,17}_SUPPORT` defines
```

- Remove support to coal < 3
- Remove support to eigenpy < 3

> - Eigen 3.4 is the minimal Eigen supported version:

Reviewer: ```suggestion
- Eigen 3.4 is now the minimal required version of Eigen:
```
